ludwig/utils/data_utils.py

```diff
@@ -827,6 +829,24 @@ def is_ludwig_ndarray_string(parm_string):
     return isinstance(parm_string, str) and parm_string[:11] == "__ndarray__"
 
 
+def get_pa_dtype(obj: Any):
+    if np.isscalar(obj):
+        return pa.from_numpy_dtype(np.array(obj).dtype)
+    elif isinstance(obj, np.ndarray) or isinstance(obj, list) or isinstance(obj, tuple):
+        return pa.list_(get_pa_dtype(obj[0]))
+    else:
+        raise ValueError(f"Unsupported type for pyarrow dtype: {type(obj)}")
+
+
+def get_pa_schema(df: DataFrame):
+    head = df.head(1)
```

Comment:
Would there be scenarios in which inferring schema from the first row may not be sufficient? E.g. in the case of `nan`s in the first row.

Reply:
Agreed, I think it might be worth taking a larger sample of rows or finding the first non-null row and use that.

Reply:
Good callout– I believe that in the vanilla Ludwig codepath, any calls to this function would only occur after NaNs are handled. That said, it is possible that to_parquet could be called in something like Predibase, so I've added a larger buffer to minimize the likelihood of failure here.